File: arrays_and_strings/2515_shortest_distance_to_target_string_in_circular_array/TS/closestTarget.ts
export {};

function closestTarget(
  words: string[],
  target: string,
  startIndex: number,
): number {
  let idxs = words.reduce((acc: number[], word: string, index: number) => {
    if (word == target) acc.push(index);
    return acc;
  }, []);
//   console.log("valid idxs:", idxs);

  if (idxs.length < 1) return -1;

  //map to calculate the steps to target idx from both directions of start index
  let steps = idxs.map((targetIdx) =>
    Math.min(
      Math.abs(targetIdx - startIndex),
      words.length - Math.abs(startIndex - targetIdx),
    ),
  );
  
//   console.log('smallest step for each idx:', steps)
  return Math.min(...steps);
}

const testCases: [string[], string, number][] = [
  [["hello", "i", "am", "leetcode", "here"], "leetcode", 3],
  [["hello", "i", "am", "leetcode", "here"], "here", 2],
  [["hello", "i", "am", "leetcode", "here"], "hello", 1],
  [["a", "b", "leetcode"], "leetcode", 0],
  [["a"], "a", 0],
  [["i", "eat", "leetcode"], "ate", 0],
];

for (const [words, target, startIndex] of testCases) {
  console.log(
    `Input:       words=${JSON.stringify(words)}, target=${JSON.stringify(target)}, startIndex=${startIndex}`,
  );
  console.log(`Output:      ${closestTarget(words, target, startIndex)}`);
  console.log("-".repeat(35));
}

File: arrays_and_strings/2515_shortest_distance_to_target_string_in_circular_array/python/closest_target.py
from typing import List


def minimum_dist(targetidx: int, startidx: int, length: int)->int:
    return min(abs(targetidx - startidx), length - abs(startidx - targetidx))

class Solution:
    def closestTarget(self, words: List[str], target: str, startIndex: int) -> int:
        valid_idxs =[]
        for idx, word in enumerate(words):
            if word == target:
                valid_idxs.append(idx)
        if not valid_idxs:
            return -1
        
        return min(map(lambda valid_idx: minimum_dist(valid_idx, startIndex, len(words)), valid_idxs))
        
                
    
    
    
if __name__ == "__main__":
    test_cases = [
        (["hello", "i", "am", "leetcode", "here"], "leetcode", 3),  # → 1
        (["hello", "i", "am", "leetcode", "here"], "here", 2),      # → 3
        (["hello", "i", "am", "leetcode", "here"], "hello", 1),     # → 1, wraps around
        (["a", "b", "leetcode"], "leetcode", 0),                    # → 1, wraps backward
        (["a"], "a", 0),                                            # → 0, single element
        (["i","eat","leetcode"], "ate", 0),
    ]

    for words, target, startIndex in test_cases:
        print(f"Input:       words={words}, target={target!r}, startIndex={startIndex}")
        print(f"Output:      {Solution().closestTarget(words, target, startIndex)}")
        print("-" * 35)